browser/default_protocol_handler_utils_win.cc

```diff
@@ -0,0 +1,505 @@
+bool CheckUserChoiceHash(base::WStringPiece protocol,
+                         base::WStringPiece user_sid) {
+  auto key_path = GetAssociationKeyPath(protocol.data());
+  if (!key_path)
+    return false;
+
+  base::win::RegKey user_choice_key(HKEY_CURRENT_USER, key_path.get(),
+                                    KEY_READ);
+  if (!user_choice_key.Valid())
+    return false;
+
+  if (user_choice_key.OpenKey(kUserChoiceKey, KEY_READ) != ERROR_SUCCESS)
+    return false;
+
+  FILETIME last_write_file_time;
+
+  // RegKey doesn't support for fetching last write time.
+  if (::RegQueryInfoKeyW(user_choice_key.Handle(), nullptr, nullptr, nullptr,
+                         nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+                         nullptr, &last_write_file_time) != ERROR_SUCCESS) {
+    return false;
+  }
+
+  SYSTEMTIME last_write_system_time;
+  if (!::FileTimeToSystemTime(&last_write_file_time, &last_write_system_time))
+    return false;
+
+  // Read ProgId
+  std::wstring prog_id;
+  if (user_choice_key.ReadValue(kProgIDKey, &prog_id) != ERROR_SUCCESS)
+    return false;
+
+  // Read Hash
+  std::wstring stored_hash;
+  if (user_choice_key.ReadValue(kHashKey, &stored_hash) != ERROR_SUCCESS)
+    return false;
+
+  auto computed_hash = GenerateUserChoiceHash(protocol, user_sid, prog_id,
+                                              last_write_system_time);
+  if (!computed_hash)
+    return false;
+
+  if (::CompareStringOrdinal(computed_hash.get(), -1, stored_hash.c_str(), -1,
+                             FALSE) != CSTR_EQUAL) {
+    return false;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+std::unique_ptr<wchar_t[]> GenerateUserChoiceHash(base::WStringPiece ext,
+                                                  base::WStringPiece sid,
+                                                  base::WStringPiece prog_id,
+                                                  SYSTEMTIME timestamp) {
+  auto user_choice = FormatUserChoiceString(ext, sid, prog_id, timestamp);
+  if (!user_choice) {
+    LOG(ERROR) << "Didn't get user choice string for generating hash.";
+    return nullptr;
+  }
+
+  return HashString(user_choice.get());
+}
+
+bool SetDefaultProtocolHandlerFor(base::WStringPiece protocol) {
+  if (IsDefaultProtocolHandlerFor(protocol)) {
+    VLOG(2) << "Already default handler for " << protocol;
+    return true;
+  }
+
+  const auto prog_id = GetBrowserProgId();
+  if (!CheckProgIDExists(prog_id)) {
+    LOG(ERROR) << "ProgId is not found - " << prog_id;
+    return false;
+  }
+
+  std::wstring user_sid;
+  if (!base::win::GetUserSidString(&user_sid)) {
+    LOG(ERROR) << "Can't get user sid";
+    return false;
+  }
+
+  if (!CheckUserChoiceHash(protocol, user_sid)) {
+    LOG(ERROR) << "UserChoice Hash mismatch";
+    return false;
+  }
+
+  // Applied FF's policy: unsupported below Windows 10 RS2
+  if (base::win::GetVersion() < base::win::Version::WIN10_RS2) {
```

Comment:
should we check this and don't show any UI if this is not supported?

Reply:
Honestly, I'm also curious why FF has this constraint for using this hash method.
When `SetDefaultProtocolHandlerFor` fails to set, system ui will be launched - https://github.com/brave/brave-core/pull/10879/files#diff-c669b2a178e5dff16197052bd67f793665e7394cc67c7bd35acb9db9d301ab76R103